Source/WebCore/loader/HistoryController.cpp

```diff
@@ -297,40 +307,50 @@ bool HistoryController::shouldStopLoadingForHistoryItem(HistoryItem& targetItem)
 
 // Main funnel for navigating to a previous location (back/forward, non-search snap-back)
 // This includes recursion to handle loading into framesets properly
-void HistoryController::goToItem(HistoryItem& targetItem, FrameLoadType type, ShouldTreatAsContinuingLoad shouldTreatAsContinuingLoad)
+void HistoryController::goToItem(HistoryItem& targetItem, FrameLoadType frameLoadType, ShouldTreatAsContinuingLoad shouldTreatAsContinuingLoad)
 {
-    LOG(History, "HistoryController %p goToItem %p type=%d", this, &targetItem, static_cast<int>(type));
+    LOG(History, "HistoryController %p goToItem %p type=%d", this, &targetItem, static_cast<int>(frameLoadType));
 
-    // shouldGoToHistoryItem is a private delegate method. This is needed to fix:
-    // <rdar://problem/3951283> can view pages from the back/forward cache that should be disallowed by Parental Controls
-    // Ultimately, history item navigations should go through the policy delegate. That's covered in:
-    // <rdar://problem/3979539> back/forward cache navigations should consult policy delegate
     RefPtr page = m_frame->page();
     if (!page)
         return;
-    if (!m_frame->loader().protectedClient()->shouldGoToHistoryItem(targetItem))
-        return;
-    if (m_defersLoading) {
-        m_deferredItem = &targetItem;
-        m_deferredFrameLoadType = type;
-        return;
-    }
 
-    // Set the BF cursor before commit, which lets the user quickly click back/forward again.
-    // - plus, it only makes sense for the top level of the operation through the frame tree,
-    // as opposed to happening for some/one of the page commits that might happen soon.
-    CheckedRef backForward = page->backForward();
-    RefPtr currentItem = backForward->currentItem(m_frame->frameID());
-    backForward->setProvisionalItem(targetItem);
-
-    // First set the provisional item of any frames that are not actually navigating.
-    // This must be done before trying to navigate the desired frame, because some
-    // navigations can commit immediately (such as about:blank).  We must be sure that
-    // all frames have provisional items set before the commit.
-    recursiveSetProvisionalItem(targetItem, currentItem.get(), ForNavigationAPI::No);
-
-    // Now that all other frames have provisional items, do the actual navigation.
-    recursiveGoToItem(targetItem, currentItem.get(), type, shouldTreatAsContinuingLoad);
+    auto finishGoToItem = [weakThis = WeakPtr { this }, frameLoadType, shouldTreatAsContinuingLoad, page, targetItem = Ref { targetItem }] (bool result) {
+        if (!weakThis)
+            return;
+        Ref protectedThis = Ref { *weakThis };
```

Comment:
I know this doesn't matter much in single threaded code but our usual pattern is:
```
RefPtr protectedThis = weakThis.get();
if (!protectedThis)
    return;
```

Which is safe even in multi-threaded code.